### PR TITLE
Add fields to BlockRecord.

### PR DESCRIPTION
**Pull request by fchirica**
Extends `BlockRecord` class to contain the two new fields `pos_ss_cc_challenge_hash` and `cc_sp_hash`. These fields will be used in checks needed by CHIP-13.

The main challenge in this PR is to ensure the block store database schema stays backward compatible, since one of its columns contain a serialized `BlockRecord` object. This patch introduces a new type, `BlockRecordDB` which matches the structure stored in the DB and converts to and from it at the `BlockStore` API boundary.

There is no need for db migration - i.e. the old DBs will still work when the clients update to this change.

### CODE REVIEW

File: chia/full_node/block_store.py
Comment (arvidn) [inline chia/full_node/block_store.py:647]: it would seem simpler to invert this condition and make it an early exit

File: chia/full_node/block_store.py
Comment (arvidn) [inline chia/full_node/block_store.py:659]: it would seem simpler to invert this condition and make it an early exit

File: chia/full_node/block_store.py
Comment (arvidn) [inline chia/full_node/block_store.py:719]: having this loop be separate looks a bit unnecessarily complex. I think it would make more sense to compute the final `BlockRecord` object in the loop where we're reading from the DB cursor directly. 
Then you wouldn't need the extra `all_blocks_plot_filters` dict either, and you can just return `ret` at the end

File: chia/full_node/block_store.py
Comment (arvidn) [inline chia/full_node/block_store.py:813]: same comment here as above. If you fold this loop into the previous loop(s), I think it will be simpler and faster

**Comment by fchirica** [comment]
Main

```
v1
6.3348s, get_block_records_by_hash
6.8179s, get_block_record
0.4246s, get_block_records_in_range
0.0042s, get_block_records_close_to_peak
9.2130s, get_block_record

v2
6.6296s, get_block_records_by_hash
6.2847s, get_block_record
0.5888s, get_block_records_in_range
0.0058s, get_block_records_close_to_peak
9.2393s, get_block_record
```

Patch
```
v1
17.6560s, get_block_records_by_hash
14.9411s, get_block_record
3.8442s, get_block_records_in_range
0.0342s, get_block_records_close_to_peak
7.8750s, get_block_record


v2
8.6854s, get_block_records_by_hash
10.6111s, get_block_record
1.7555s, get_block_records_in_range
0.0156s, get_block_records_close_to_peak
8.4783s, get_block_record
```

File: chia/full_node/block_store.py
Comment (arvidn) [inline chia/full_node/block_store.py:716]: I imagine combining these queries into one provide a speed-up compared to making two separate queries (as you do above). Is there a reason to not do the same in `get_block_record()`? We probably don't care enough about v1 performance.

**Comment by arvidn** [comment]
This looks good to me. the performance impact is a bit greater than I expected, but I don't think there's any obvious simple fix to that. We'll have to optimize this more over time.

I will approve it once we have one more review.

**Review by xdustinface — CHANGES_REQUESTED** [review]
The added test needs little more love imo and i wonder if there is no other way to solve this which doesn't involve affecting all the block store queries and which doesn't lead to duplication (split into memory/DB classes) of the `BlockRecord` data structure given we only seem to be interested in the this data for the last few blocks.

I didn't look a lot into the CHIP13 PR yet but after a quick check it seems like this PR is only to make this here possible https://github.com/Chia-Network/chia-blockchain/blob/1e40152e484ee1ff1d6f3e63b3c5f0997e602ea6/chia/consensus/block_header_validation.py#L495-L513

So we could maybe just maintain a `PlotFilterInfo` cache for the last `UNIQUE_PLOTS_WINDOW` or whatever safe amount of blocks to solve this with less side effects? 

Maybe i miss something but i wanted to throw out the idea, let me know what you think!

File: tests/core/full_node/stores/test_block_store.py
Comment (fchirica) [reply tests/core/full_node/stores/test_block_store.py:377]: I'm keeping consistency with the above tests.

File: tests/core/full_node/stores/test_block_store.py
Comment (fchirica) [reply tests/core/full_node/stores/test_block_store.py:378]: This is again the pattern used by the above tests. From what I can see, one store is needed by blockchain DB and one is the one that we're doing tests on. I think we need two, so we can do operations on block store and on the blockchain independently.

File: tests/core/full_node/stores/test_block_store.py
Comment (fchirica) [reply tests/core/full_node/stores/test_block_store.py:385]: `store_2` is not thrown away, it'll be used by `Blockchain` object.

File: tests/core/full_node/stores/test_block_store.py
Comment (fchirica) [reply tests/core/full_node/stores/test_block_store.py:374]: My logic was we want to hit this case `block.reward_chain_block.challenge_chain_sp_vdf is None`, and it's rarer and might not be present in a small (50 blocks) blockchain. I thought this feature is quite critical, so more testing is better. Tho probably 400 will do the job, but for 50 I'm not so sure.

**Comment by fchirica** [comment]
> The added test needs little more love imo and i wonder if there is no other way to solve this which doesn't involve affecting all the block store queries and which doesn't lead to duplication (split into memory/DB classes) of the `BlockRecord` data structure given we only seem to be interested in the this data for the last few blocks.
> 
> I didn't look a lot into the CHIP13 PR yet but after a quick check it seems like this PR is only to make this here possible
> 
> https://github.com/Chia-Network/chia-blockchain/blob/1e40152e484ee1ff1d6f3e63b3c5f0997e602ea6/chia/consensus/block_header_validation.py#L495-L513
> 
> So we could maybe just maintain a `PlotFilterInfo` cache for the last `UNIQUE_PLOTS_WINDOW` or whatever safe amount of blocks to solve this with less side effects?
> 
> Maybe i miss something but i wanted to throw out the idea, let me know what you think!

This is already cached most of the time: https://github.com/Chia-Network/chia-blockchain/blob/main/chia/consensus/blockchain.py#L688 I think this dict keeps a lot of recent `BlockRecord` in memory, so the DB won't be involved at all (we obtain it from the full block: https://github.com/Chia-Network/chia-blockchain/blob/main/chia/consensus/full_block_to_block_record.py). The problem is, when we need more than this cache (which I think is quite big), we'll go back to DB queries. So, while they're usually not used, in case of a very deep reorg, we need this functionality as a fallback or the already existing cache.

File: tests/core/full_node/stores/test_block_store.py
Comment (xdustinface) [reply tests/core/full_node/stores/test_block_store.py:385]: But `store_2` is not created here? The here linked line calls `BlockStore.create` and throws away (doesn't assign it) the created instance.

File: tests/core/full_node/stores/test_block_store.py
Comment (xdustinface) [reply tests/core/full_node/stores/test_block_store.py:377]: Not sure if other tests doing something unnecessary justifies continuing doing it? :) If this is something we want to test it should be some separate test which tests this rather imo since its a global thing and there is no point to copy/paste it into all tests.

File: tests/core/full_node/stores/test_block_store.py
Comment (xdustinface) [reply tests/core/full_node/stores/test_block_store.py:374]: Okay yeah, so we probably should know how many blocks this case requires and take enough to cover it.

File: tests/core/full_node/stores/test_block_store.py
Comment (xdustinface) [reply tests/core/full_node/stores/test_block_store.py:378]: To me it seems like you can only use one `DBWrapper` here, drop all `_2` prefixes, drop the `store = await BlockStore.create(db_wrapper)` and the test then still does what its supposed to do?

**Review by xdustinface — APPROVED** [review]

**Review by arvidn — APPROVED** [review]

**Pull request merged**
